### PR TITLE
Better high level C++ documentation

### PR DESCRIPTION
**Pull request by goldsborough**
I wrote some high level docs for the larger PyTorch C++ universe and the C++ frontend specifically. Happy for reviews, but let's please also land this ASAP so I can point users at something that looks more ready baked than the C++ docs landing page (https://pytorch.org/cppdocs) does right now.

@ezyang @soumith 

CC @ebetica

### CODE REVIEW

File: docs/cpp/source/frontend.rst [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+To see more complete examples of using the PyTorch C++ frontend, please do
+explore `the example repository
+<https://github.com/goldsborough/examples/tree/cpp/cpp>`_.
+
+Philosophy and Origin
+---------------------
+
+The PyTorch C++ frontend was originally born out of the necessity to train
```
Comment (ezyang) [inline docs/cpp/source/frontend.rst]: I want to copy-edit this section.

File: docs/cpp/source/index.rst [hunk 2]
```diff
@@ -1,36 +1,180 @@
 PyTorch C++ API
 ===============
 
-The PyTorch C++ API is a research and production ready C++ interface to PyTorch,
-a library for tensors and dynamic neural networks with strong GPU acceleration.
+These pages provide documentation for the public portions of the PyTorch C++
```
Comment (ezyang) [inline docs/cpp/source/index.rst]: This intro can be just one sentence.

**Review by ezyang — APPROVED** [review]
I want to do some copyediting, but if you really want to land this right now, OK.

File: docs/cpp/source/frontend.rst [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+      if (batch_index++ % 10 == 0) {
+        std::cout << "Epoch: " << epoch << " | Batch: " << batch_index
+                  << " | Loss: " << loss << std::endl;
+        // Serialize your model periodically as a checkpoint.
+        torch::save(net, "net.pt");
+      }
+    }
+
+To see more complete examples of using the PyTorch C++ frontend, see `the example repository
+<https://github.com/goldsborough/examples/tree/cpp/cpp>`_.
```
Comment (ezyang) [inline docs/cpp/source/frontend.rst]: We should move this repo to the pytorch org.

File: docs/cpp/source/frontend.rst [hunk 4]
```diff
@@ -101,50 +101,38 @@ To see more complete examples of using the PyTorch C++ frontend, see `the exampl
...
-(especially any kind of numeric operation). Translating your model to C++ will
-usually not make it magically faster, and you may often not notice any
-significant performance gain at all! The problem the C++ frontend solves is not
-a performance problem. The problem it solves is being able to write friendly,
-flexible and intuitive machine learning applications in environments where C++
-is your language of choice, or necessity. If you would prefer to write Python,
+PyTorch's C++ frontend was designed with the idea that the Python frontend
+is great, and should be used when possible; but in some settings, performance
+and portability requirements make the use of the Python interpreter infeasible.
+For example, Python is poor choice for low latency, high performance or
```
Comment (goldsborough) [inline docs/cpp/source/frontend.rst]: a

File: docs/cpp/source/installing.rst [hunk 5]
```diff
@@ -1,7 +1,7 @@
 Installing C++ Distributions of PyTorch
 =======================================
 
-We provide ready baked distributions of all headers, libraries and CMake
+We provide binary distributions of all headers, libraries and CMake
```
Comment (goldsborough) [inline docs/cpp/source/installing.rst]: Will I find "ready baked" in the thesaurus entry for "binary"? 😛

File: docs/cpp/source/frontend.rst (hunk 3 above)
Comment (goldsborough) [reply docs/cpp/source/frontend.rst]: It's too unpolished right now

Comment (goldsborough) [reply docs/cpp/source/frontend.rst]: But ultimately yes

**Pull request closed without merging**